test/specs/commonTests/stylesFunction-test.tsx

```diff
@@ -0,0 +1,115 @@
+import * as React from 'react'
+import * as _ from 'lodash'
+import { UIComponent } from 'src/lib'
+import { Extendable } from 'types/utils'
+import { ICSSInJSStyle } from 'types/theme'
+import { getTestingRenderedComponent } from 'test/utils'
+
+type AttrValue = 'props' | 'state'
+
+interface IProps {
+  propsAttr?: AttrValue
+  commonAttr?: AttrValue
+}
+
+interface IState {
+  commonAttr?: AttrValue
+  stateAttr?: AttrValue
+}
+
+type IPropsAndState = IProps & IState
+
+const testClassName = 'ui-test-component'
+
+const testStylesForComponent = ({
+  props,
+  state,
+  expected,
+}: { props?: IProps; state?: IState; expected?: IPropsAndState } = {}) => {
+  class TestComponent extends UIComponent<Extendable<IProps>, IState> {
+    public static className = testClassName
+    public static handledProps = ['propsAttr', 'commonAttr', 'styles']
+
+    public state = state
+
+    public renderComponent({ ElementType, classes, rest }): React.ReactNode {
+      return <ElementType {...rest} className={classes.root} />
+    }
+  }
+
+  const TestStylesComponent = (props: Extendable<IProps>) => (
+    <TestComponent
+      {...props}
+      styles={{
+        root: ({ props }: { props: IPropsAndState }): ICSSInJSStyle => {
+          expect(_.mapValues(expected, (val, key) => props[key])).toEqual(expected)
+          return {}
+        },
+      }}
+    />
+  )
+
+  getTestingRenderedComponent(TestStylesComponent, <TestStylesComponent {...props} />)
+}
+
+describe('styles function receives as parameter', () => {
+  it('props only if state is not set', () => {
+    testStylesForComponent({ expected: {} })
+
+    testStylesForComponent({
+      props: { propsAttr: 'props' },
+      expected: { propsAttr: 'props' },
+    })
+
+    testStylesForComponent({
+      props: { propsAttr: 'props', commonAttr: 'props' },
+      expected: { propsAttr: 'props', commonAttr: 'props' },
+    })
+  })
+
+  it('simple merge of props and state if both are set but there are no overlapping properties', () => {
```

Comment:
it would be better to describe result in the test name, like `props and state properties are merged together` and `props properties take precedence over ones of state` for the next one

Reply:
I don't think I understand, let's discuss

Reply:
discussed with @kuzhelov and amended